File: src/main/java/net/adventuregame/resources/ResourceLocation.java
package net.adventuregame.resources;

public record ResourceLocation(String namespace, String path) {

    public ResourceLocation {
        if (!namespace.matches("[a-z0-9_.-]+") || !path.matches("[a-z0-9_/.-]+")) {
            throw new IllegalArgumentException("Invalid ResourceLocation: " + namespace + ":" + path);
        }
    }

    public static ResourceLocation withDefaultNamespace(String path) {
        return new ResourceLocation("adventuregame", path);
    }

    public static ResourceLocation fromString(String id) {
        String[] parts = id.split(":", 2);
        if (parts.length == 1) return new ResourceLocation("adventuregame", parts[0]);
        return new ResourceLocation(parts[0], parts[1]);
    }

    @Override
    public String toString() {
        return namespace + ":" + path;
    }
}

File: src/main/kotlin/net/adventuregame/resources/ResourceLocation.kt
package net.adventuregame.resources

@JvmRecord
data class ResourceLocation(val namespace: String?, val path: String?) {
    override fun toString(): String {
        return "$namespace:$path"
    }

    init {
        require(!(!namespace!!.matches("[a-z0-9_.-]+".toRegex()) || !path!!.matches("[a-z0-9_/.-]+".toRegex()))) { "Invalid ResourceLocation: " + namespace + ":" + path }
    }

    companion object {
        fun withDefaultNamespace(path: String?): ResourceLocation {
            return ResourceLocation("adventuregame", path)
        }

        fun fromString(id: String): ResourceLocation {
            val parts: Array<String?> = id.split(":".toRegex(), limit = 2).toTypedArray()
            if (parts.size == 1) return ResourceLocation("adventuregame", parts[0])
            return ResourceLocation(parts[0], parts[1])
        }
    }
}

